### PR TITLE
CO-1972 form process inside a job

### PR DESCRIPTION
**Pull request by quentingigon**

### CODE REVIEW

File: sms_sponsorship/forms/sms_sponsorship_registration_form.py [hunk 1]
```diff
@@ -171,7 +173,7 @@ def form_before_create_or_update(self, values, extra_values):
             partner_vals = self._get_partner_vals(extra_values)
             if partner:
                 # update existing partner
-                partner.write(partner_vals)
+                self.with_delay().update_partner(partner_vals)
```
Comment (ecino) [inline sms_sponsorship/forms/sms_sponsorship_registration_form.py]: I think it would be best to put the update_partner method inside the res.partner object so that you can do `partner.with_delay().update_partner(partner_vals)`

File: sms_sponsorship/forms/sms_sponsorship_registration_form.py [hunk 2]
```diff
@@ -186,11 +188,22 @@ def form_before_create_or_update(self, values, extra_values):
                     })
             sponsorship.on_change_group_id()
 
-        def form_after_create_or_update(self, values, extra_values):
+        @job(default_channel="root.res_partner")
```
Comment (ecino) [inline sms_sponsorship/forms/sms_sponsorship_registration_form.py]: Channel should be 'root.sms_sponsorship'

File: sms_sponsorship/forms/sms_sponsorship_registration_form.py [hunk 3]
```diff
@@ -186,11 +188,22 @@ def form_before_create_or_update(self, values, extra_values):
...
+        @related_action(action='related_action_update_partner')
+        def update_partner(self, partner_vals):
+            self.env['res.partner'].search([
```
Comment (ecino) [inline sms_sponsorship/forms/sms_sponsorship_registration_form.py]: If you put this method inside res.partner object, you can simply do `return self.write(partner_vals)`

File: sms_sponsorship/forms/sms_sponsorship_registration_form.py [hunk 4]
```diff
@@ -186,11 +188,22 @@ def form_before_create_or_update(self, values, extra_values):
...
+        @job(default_channel="root.recurring_contract")
+        @related_action(action='related_action_finalize_form')
+        def finalize_form(self, sponsorship):
             # validate sponsorship and send confirmation email
-            sponsorship = self.main_object.sudo()
+            # we use partner name temporarily for testing reasons, should
+            # use sponsorship.id
             sms_request = self.env['sms.child.request'].sudo().search([
-                ('sponsorship_id', '=', self.main_object.id)
+                ('partner_id.name', '=', sponsorship.partner_id.name)
```
Comment (ecino) [inline sms_sponsorship/forms/sms_sponsorship_registration_form.py]: No, the previous search was actually better.

File: sms_sponsorship/forms/sms_sponsorship_registration_form.py [hunk 5]
```diff
@@ -220,6 +233,9 @@ def form_after_create_or_update(self, values, extra_values):
             sms_request.complete_step2()
             sponsorship.button_generate_invoices()
 
+        def form_after_create_or_update(self, values, extra_values):
+            self.with_delay().finalize_form(self.main_object.sudo())
```
Comment (ecino) [inline sms_sponsorship/forms/sms_sponsorship_registration_form.py]: Again, I think it makes sense to move the job function inside the `recurring.contract` object so that you can call here `self.main_object.sudo().with_delay().finalize_form()`

**Review by ecino — CHANGES_REQUESTED** [review]

File: sms_sponsorship/forms/sms_sponsorship_registration_form.py (hunk 4 above)
Comment (quentingigon) [reply sms_sponsorship/forms/sms_sponsorship_registration_form.py]: I know, it was just not working on tests because I can't create a sms_request with an attributed sponsorship so I couldn't find a sms_request.

File: sms_sponsorship/forms/sms_sponsorship_registration_form.py [hunk 6]
```diff
@@ -162,7 +162,7 @@ def _form_validate_alpha_field(self, field, value):
 
         # Form submission
         #################
-        def form_before_create_or_update(self, values, extra_values):
+        def form_before_create_or_update(self, *values, **extra_values):
```
Comment (ecino) [inline sms_sponsorship/forms/sms_sponsorship_registration_form.py]: Why did you change that?

File: sms_sponsorship/models/res_partner.py [hunk 7]
```diff
@@ -58,3 +59,8 @@ def sms_send_step2_confirmation(self, child_request):
             content_subtype='html'
         )
         return True
+
+    @job(default_channel="root.res_partner")
+    @related_action(action='related_action_update_partner')
+    def update_partner(self, partner_vals):
+        self.write(partner_vals)
```
Comment (ecino) [inline sms_sponsorship/models/res_partner.py]: Should return the result.

File: sms_sponsorship/models/recurring_contract.py [hunk 8]
```diff
@@ -59,3 +71,38 @@ def create_sms_sponsorship(self, vals, partner, sms_child_request):
...
+            self.signal_workflow('contract_validated')
+
+        # if sponsor directly payed
+        if pay_first_month_ebanking:
+            # load payment view ? TODO
+            _logger.error("Activate sponsorship is not yet implemented")
+
+        # update sms request
+        sms_request.complete_step2()
+        self.button_generate_invoices()
```
Comment (ecino) [inline sms_sponsorship/models/recurring_contract.py:108]: Should return True

**Review by ecino — CHANGES_REQUESTED** [review]

File: sms_sponsorship/forms/sms_sponsorship_registration_form.py (hunk 6 above)
Comment (quentingigon) [reply sms_sponsorship/forms/sms_sponsorship_registration_form.py]: I didn't change that in my commit, must have been a previous error

**Review by ecino — APPROVED** [review]
